feat: lexpol integration

yhru opened the pull request:
This pull request introduces several changes focused on integrating Lexpol functionality into the application. The changes include new components, updates to existing components, and modifications to controllers and GraphQL schema to support Lexpol operations.

### Lexpol Integration:

* **New Lexpol Component**:
  - [`app/components/editable_champ/lexpol_component.rb`](diffhunk://#diff-00548faf1512d54902799d1bb2bdf4b3651fd1d2ff6b7862071f3f2f119726cdR1-R2): Added a new component for Lexpol.
  - [`app/components/editable_champ/lexpol_component/lexpol_component.html.haml`](diffhunk://#diff-5e031281897342f4ffe592d5132667c4de8862929c1a8e65c6678efd5de68cedR1-R19): Added HTML structure for displaying Lexpol data.

* **Lexpol Models and API Integration**:
  - [`app/components/types_de_champ_editor/champ_component.rb`](diffhunk://#diff-9f1cc48aaaf637e4b11d7bf3fb928b0dc93af36f7810d1954e75ef7850011f06R11-R22): Introduced `lexpol_models` method to fetch models from Lexpol API.
  - [`app/models/champs/lexpol_champ.rb`](diffhunk://#diff-a0b0929105ec79630be52f019c183e65f118af7a3c64d062c65b5744beabc320R1-R16): Created a new model for LexpolChamp with methods to generate or update dossiers.

* **Controller Updates**:
  - [`app/controllers/administrateurs/procedures_controller.rb`](diffhunk://#diff-d488a8cef7849b02ae0321f902c6dccc6c4ce45743ba3eff31114081859cbac0R592-R602): Added `update_lexpol` method to handle Lexpol dossier updates.
  - [`app/controllers/instructeurs/lexpol_controller.rb`](diffhunk://#diff-5a7519dfcc387c6a83e5f4ba57fb97ce0a1fbf5e25a5f98949a8d7968bff3b30R1-R28): Created a new controller for handling Lexpol-related actions.

* **GraphQL Schema Updates**:
  - [`app/graphql/schema.graphql`](diffhunk://#diff-cf23d894acfdcdc84db608293864fcd2faa312864bb85345f330fb3d55c30d1cR3200-R3266): Added new types and descriptors for LexpolChamp and updated the `TypeDeChamp` enum. [[1]](diffhunk://#diff-cf23d894acfdcdc84db608293864fcd2faa312864bb85345f330fb3d55c30d1cR3200-R3266) [[2]](diffhunk://#diff-cf23d894acfdcdc84db608293864fcd2faa312864bb85345f330fb3d55c30d1cR4813-R4817)
  - [`app/graphql/types/champs/lexpol_champ_type.rb`](diffhunk://#diff-d848865c2e68aad23e61c400bd4c84527e27907c575df17bc6b57998b380992bR1-R9): Defined GraphQL type for LexpolChamp.

These changes collectively enhance the application's capability to handle Lexpol dossiers, providing new components, API integrations, and necessary schema updates.

yhru committed:
feat: lexpol integration

yhru committed:
feat: new line

yhru committed:
feat: generate graphql idl schema

yhru committed:
feat: rake job for feature_flag lexpol

yhru committed:
feat: add test and allow create to send var

yhru committed:
feat: use store accessor instead of data

maatinito commented on app/components/types_de_champ_editor/champ_component/champ_component.html.haml:
```diff
@@ -64,8 +64,25 @@
+                                { class: 'fr-select small-margin small width-100', id: dom_id(type_de_champ, :lexpol_modele) }
+                - else
+                  .fr-message.fr-message--error
+                    %p Aucune modèle disponible ou erreur lors de la récupération des modèles.
```
Aucune modèle ?

maatinito commented on app/components/types_de_champ_editor/champ_component/champ_component.html.haml:
```diff
@@ -64,8 +64,25 @@
+            - if type_de_champ.lexpol?
+              .cell.fr-mt-1w
+                = form.label :lexpol_modele, "Sélectionner un modèle Lexpol", class: 'flex-grow', for: dom_id(type_de_champ, :lexpol_modele)
+                - if lexpol_models.any?
```
Les modèles devraient être triée ET tronqués à 40 caractères

yhru replied:
Pour le tri, on le fait par ordre alphabétique ?

maatinito replied:
est-ce que tu verrais un ordre différent ?

yhru replied:
Actuellement il me semble que c'est dans l'ordre antéchronologique

Je ne sais pas comment les utilisateurs finaux vont se servir de cette liste si tu préfères alphabétique à celui en place je peux changer

maatinito commented on app/controllers/instructeurs/dossiers_controller.rb:
```diff
@@ -371,6 +371,55 @@ def pieces_jointes
+      puts "Procedure associée au dossier : #{dossier.procedure.inspect}"
+      puts "Procedure attendue : #{params[:procedure_id]}"
+
+      mapping = (champ.type_de_champ.lexpol_mapping || "")
```
je pensais que l'on mettait un mapping par ligne (\n) et non des virgules

maatinito commented on app/controllers/instructeurs/dossiers_controller.rb:
```diff
@@ -371,6 +371,55 @@ def pieces_jointes
+        .map { |pair| pair.split('=').map(&:strip) }
+        .to_h
+
+      variables = dossier.champs.each_with_object({}) do |champ, hash|
```
je comprends l'interet d'envoyer toutes les variables (par défaut pas de configuration)
En meme temps: 
* il faut aussi envoyer les annotations 
* il faut pouvoir indiquer que la variable lexpol toto.tutu est égal au champ Numéro_DN.date_de_naissance ou que la variable lexpol demande.date est égal à la date de dépot du dossier (ce n'est pas un champ)
* le code devrait etre factorisé dans lexpol api

maatinito commented on app/controllers/instructeurs/dossiers_controller.rb:
```diff
@@ -371,6 +371,55 @@ def pieces_jointes
         .filter { _1.class.in?([Champs::PieceJustificativeChamp, Champs::TitreIdentiteChamp]) }
     end
 
+    def create_lexpol_dossier
```
pourquoi cela semble faire doublon avec le lexpol controller ?

maatinito commented on app/models/concerns/lexpol_concern.rb:
```diff
@@ -0,0 +1,44 @@
+  def refresh_lexpol_data!
+    return if value.blank?
+
+    lexpol_service = APILexpol.new
```
un peu dommage de réinitialiser un lexpol api alors qu'il est créé juste au dessus ? 
est-ce que cela va demander un nouveau token ?

maatinito commented on app/views/shared/dossiers/_edit_annotations.html.haml:
```diff
@@ -6,5 +6,26 @@
+            %strong Statut du dossier :
+            = champ_lexpol.data&.[]('lexpol_status') || 'Inconnu'
+
+          - if champ_lexpol.data&.[]('lexpol_dossier_url').present?
```
je pense que ca devrait etre un lien sur le numéro NOR plutot qu'un bouton : ce n'est pas une action mais bien un lien vers une autre page

maatinito commented on app/services/api_lexpol.rb:
```diff
@@ -0,0 +1,99 @@
+    @credentials = {
+      email: ENV.fetch('API_LEXPOL_EMAIL'),
+      motdepasse: ENV.fetch('API_LEXPOL_PASSWORD'),
+      email_agent: ENV.fetch('API_LEXPOL_AGENT_EMAIL')
```
Le mail de l'agent : c'est le mail sous lequel tu es connecté en ce moment... 
Chaque fois qu'un agent va travailler dans mes-demarches, il faut qu'il le fasse en son nom. 
Donc, le initialize devrait prendre à chaque fois le user courant. 
Pb : notre mail n'est pas enregistré sur lexpol pour l'instant. Faut que je vois comment faire mais à priori je dirais que tu peux patcher en disant si le mail du user courant finit en .gov.pf alors c'est ce mail qu'il faut utiliser sinon prendre API_LEXPOL_AGENT_EMAIL.
Et du coup, ca me fait penser qu'un user peut ne pas avoir le droit de manipuler la partie lexpol et il faudra que Mes-Démarches grise les actions sur le dossier...

yhru committed:
chore: change according to review 1

yhru committed:
chore: change according to review 2

yhru committed:
feat: allow to use all available modeles

yhru committed:
feat: unified create and update logic to an upsert

yhru committed:
refactor: put lexpol_controller logic into lexpol_service

yhru committed:
feat: chain var and send annotations

yhru committed:
feat: send metadata and fix lexpol controller spec

yhru committed:
fix: send VAR for spec

yhru committed:
feat: dossier_link and repetition field

yhru committed:
fix: lexpol status

yhru committed:
feat: format for date datetime table and textarea (markdown to html)

yhru committed:
chore: change according to review

maatinito committed:
By default, send to lexpol all champs
Handle the syntax 'variable' without '=' sign

maatinito committed:
handle "repetition.count' syntax
handle "repetition.champ' synta
object_field_value no longer format date and time. Formating is dedicated to format_lexpol_value
long format for date, datetime (5 mai 1970 à 19:23)
table generation correction to fix pdf conversion in Lexpol
markdown generation correction to fix pdf conversion in Lexpol.

yhru committed:
fix: preview with lexpol field

yhru committed:
feat: amelioration

yhru committed:
feat: refresh when file is cancel

maatinito committed:
Let lexpol controller becomes a ChampController

maatinito committed:
First day of month is now displayed as "1er" instead of "1"

maatinito committed:
enhanced default variables (Dossier déposé le, ...)

maatinito committed:
apilexpol sends current user email instead of fixed email
For super admin, apilexpol look for current service siret in LEXPOL_SERVICE_EMAILS to find email to use on Lexpol
